File: Tree/Recursive Traversal/Distribute Candies/postOrder_traversal.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complextiy :- O(n)
// Space complexity :- O(n)

// Approach :-  
// Start with bottom-up approach if the child have extra coins it will pass it to parent
// if child is passing a coin or demanding n number of coins from parent 
// then that will be counted as a move 
// we can return the number of coins from every node needed/extra to it's parent


// Link :- https://www.geeksforgeeks.org/problems/distribute-candies-in-a-binary-tree/1

class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int x) {
        data = x;
        left = right = nullptr;
    }
};

class Solution {
  private:
    int solve(Node* node, int &res){
        if(!node)       return 0;
        
        int leftVal = solve(node -> left, res);
        int rightVal = solve(node -> right, res);
        
        int coinsNeed = leftVal + rightVal + node -> data - 1;
        res += abs(coinsNeed);
        
        return coinsNeed;
    }
  
  public:
    int distCandy(Node* root) {
        // code here
        int res = 0;
        solve(root, res);
        return res;
    }
};

File: Tree/Recursive Traversal/Distribute Candies/postOrder_traversal.java
import java.util.*

// Time complexity :- O(n)
// Space complexity :- O(n)

// Approach :- 
// we've to calculate how much movement of candies have to be done in complete tree 
// we'll use post order traversal for that 
// we keep track of all the movement in a variable which will be the sum of subtree candies 
// current node will keep 1 candy and rest it will pass to the parent 

// Code :- 

class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

class Solution {
    private int res;
    
    private int solve(Node node){
        // base case
        if(node == null)    
            return 0;
        
        // find the coin the have to move to from left subtree 
        int coinFromLeft = solve(node.left);
        // find the coin the have to move to from right subtree 
        int coinFromRight = solve(node.right);
        
        // total movement will be sum from both subtree 
        res += Math.abs(coinFromLeft) + Math.abs(coinFromRight);
        
        // current node will keep 1 coins and rest all it will pass to it parent node
        return (node.data + coinFromLeft + coinFromRight - 1);
    }
    
    public int distCandy(Node root) {
        res = 0;
        solve(root);
        return res;
    }
}